jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -1530,26 +1551,127 @@ Method getMethodHelper(
+				Class<?> candidateRetType = bestCandidate.getReturnType();
+				Class<?> resultRetType = result.getReturnType();
+				if ((candidateRetType != resultRetType) && candidateRetType.isAssignableFrom(resultRetType)) {
+					bestCandidate = result;
+				}
+			}
 		}
-		boolean publicMethod = ((result.getModifiers() & Modifier.PUBLIC) != 0);
-		if ((methodList != null) && publicMethod) {
-			methodList.add(result);
+	}
+	return cacheMethod(bestCandidate);
+}
+
+/**
+ * Helper method searches all interfaces implemented by superclasses from the top to the bottom
+ * for the most specific method declared in one of these interfaces.
+ *
+ * @param currentClass the class to be searched, including the current class and all superclasses
+ * @param name the specified method's name
+ * @param strSig the string of the specified method's signature
+ * @param parameterTypes the types of the arguments of the specified method
+ * @return the most specific method selected from all interfaces from each superclass of the current class;
+ *         otherwise, return the method of the first interface from the top superclass
+ *         if the return types of all specified methods are identical.
+ */
+private static Method getMostSpecificMethodFromAllInterfacesOfAllSuperclasses(Class<?> currentClass, String name, String strSig, Class<?>... parameterTypes) {
+	Method candidateMethod = null;
+	if (currentClass != Object.class) {
+		/* get to the top superclass first. if all return types end up being the same the interfaces from this superclass have priority. */
+		candidateMethod = getMostSpecificMethodFromAllInterfacesOfAllSuperclasses(currentClass.getSuperclass(), name, strSig, parameterTypes);
+		
+		/* search all interfaces of current class, comparing against result from previous superclass. */
+		candidateMethod = getMostSpecificMethodFromAllInterfacesOfCurrentClass(currentClass, candidateMethod, name, strSig, parameterTypes);
+	}
+	return candidateMethod;
+}
+
+/**
+ * Helper method searches all interfaces implemented by the current class or interface 
+ * for the most specific method declared in one of these interfaces.
+ *
+ * @param currentClass the class or interface to be searched
+ * @param potentialCandidate potential candidate from superclass, null if currentClass is an interface
+ * @param name the specified method's name
+ * @param strSig the string of the specified method's signature
+ * @param parameterTypes the types of the arguments of the specified method
+ * @return the most specific method selected from all interfaces;
+ *         otherwise if return types from all qualifying methods are identical, return an arbitrary method.
+ */
+private static Method getMostSpecificMethodFromAllInterfacesOfCurrentClass(Class<?> currentClass, Method potentialCandidate, String name, String strSig, Class<?>... parameterTypes) {
+	Method bestMethod = potentialCandidate;
+	Method[] methodCandidates = currentClass.getMethods();
```

Comment:
`getMethods()` is going to be invoked for the superclasses from the top to the bottom, and the later invocations include the results of previous calls which seems an overkill.
can we modify `getMethodImpl(name, parameterTypes, strSig)` to introduce a `startingPoint`, particularly for the case that the method found is declared by an interface class?

Reply:
I agree that it is overkill.
I think it would still be useful to take advantage of the `getMethods` helper functions for the sake of not duplicating code, and combining the logic was helpful in that I was able to improve correctness of `getMethod` and `getMethods` at the same time. I can limit the superclasses searches for this particular case.

Reply:
Not against the idea. `getMethods()` just is a heavyweight api, can `getMethodImpl` & `getDeclaredMethodImpl` be used for the superclasses from the top to the bottom instead?

> improve correctness of getMethod and getMethods at the same time

`getMethod()` calls `getMethodHelper()` which is being fixed by this PR.
does `getMethods()` miss any method?

Reply:
> Not against the idea. `getMethods()` just is a heavyweight api, can `getMethodImpl` & `getDeclaredMethodImpl` be used for the superclasses from the top to the bottom instead?

I think your idea of adding a `startingPoint` type variable to `getMethodImpl` would be a good solution as well.

> > improve correctness of getMethod and getMethods at the same time
> 
> `getMethod()` calls `getMethodHelper()` which is being fixed by this PR.
> does `getMethods()` miss any method?

There is one case for `getMethods` being fixed here as well that was uncovered by one of the tests from #7623. Maybe that will explain more how that line of thinking came about too :) I will make that more obvious in the issue description.

Reply:
@JasonFengJ9  I made some updates to this approach to tighten things up. `getMostSpecificMethodFromAllInterfacesOfCurrentClass` will now call `getMethodSet` directly, and then only search for interfaces and not classes. Let me know what you think.